public/app/plugins/datasource/loki/metricTimeSplit.ts

```diff
@@ -0,0 +1,73 @@
+// every timestamp in this file is a number which contains an unix-timestamp-in-millisecond format,
+// like returned by `new Date().getTime()`. this is needed because the "math"
+// has to be done on integer numbers.
+
+import { chunk } from 'lodash';
+
+// we are trying to be compatible with
+// https://github.com/grafana/loki/blob/089ec1b05f5ec15a8851d0e8230153e0eeb4dcec/pkg/querier/queryrange/split_by_interval.go#L327-L336
+
+function expandTimeRange(startTime: number, endTime: number, step: number): [number, number] {
+  // startTime is decreased to the closes multiple-of-step, if necessary
+  const newStartTime = startTime - (startTime % step);
+
+  // endTime is increased to the closed multiple-of-step, if necessary
+  let newEndTime = endTime;
+  const endStepMod = endTime % step;
+  if (endStepMod !== 0) {
+    newEndTime += step - endStepMod;
+  }
+
+  return [newStartTime, newEndTime];
+}
+
+function getAllPoints(startTime: number, endTime: number, step: number): number[] {
+  const result: number[] = [];
+  // we get start&end times that are nicely aligned to `step`, meaning the distance
+  // between them is an integer multiple of `step`
+  const [alignedSart, alignedEnd] = expandTimeRange(startTime, endTime, step);
+  for (let current = alignedSart; current <= alignedEnd; current += step) {
+    result.push(current);
+  }
+  return result;
+}
+
+export function getRanges(
+  startTime: number,
+  endTime: number,
+  step: number,
+  idealRangeSize: number
```

Comment:
What do you think about renaming this to something like `idealRangeDuration` or `idealRangeMs`. I was having difficulties understanding it until I figured out it's actually not the total size of ranges we want, but the "size"/duration of a single range.